tests/system/Validation/ValidationTest.php

```diff
@@ -87,6 +89,107 @@ public function testSetRulesStoresRules()
         $this->assertEquals($rules, $this->validation->getRules());
     }
 
+    public function testSetRuleStoresRule()
+    {
+        $this->validation->setRules([]);
+        $this->validation->setRule('foo', null, 'bar|baz');
+
+        $this->assertSame([
+            'foo' => [
+                'label' => null,
+                'rules' => 'bar|baz',
+            ],
+        ], $this->validation->getRules());
+    }
+
+    public function testSetRuleAddsRule()
+    {
+        $this->validation->setRules([
+            'bar' => [
+                'label' => null,
+                'rules' => 'bar|baz',
+            ],
+        ]);
+        $this->validation->setRule('foo', null, 'foo|foz');
+
+        $this->assertSame([
+            'foo' => [
+                'label' => null,
+                'rules' => 'foo|foz',
+            ],
+            'bar' => [
+                'label' => null,
+                'rules' => 'bar|baz',
+            ],
+        ], $this->validation->getRules());
+    }
+
+    public function testSetRuleOverwritesRule()
+    {
+        $this->validation->setRules([
+            'foo' => [
+                'label' => null,
+                'rules' => 'bar|baz',
+            ],
+        ]);
+        $this->validation->setRule('foo', null, 'foo|foz');
+
+        $this->assertSame([
+            'foo' => [
+                'label' => null,
+                'rules' => 'foo|foz',
+            ],
+        ], $this->validation->getRules());
+    }
+
+    /**
+     * @dataProvider setRuleRulesFormatCaseProvider
+     */
+    public function testSetRuleRulesFormat(bool $expected, $rules): void
+    {
+        if (! $expected) {
+            $this->expectException(TypeError::class);
+            $this->expectExceptionMessage('$rules must be of type string|array');
+        } else {
+            $this->addToAssertionCount(1);
```

Comment:
This should be after the validation call, just to be sure the call completes successfully. Execution should stop with the exception so the `else` will be irrelevant there.

Reply:
jup, thats right